python/ffsim/hamiltonians/diagonal_coulomb_hamiltonian.py

```diff
@@ -0,0 +1,213 @@
+        return op
+
+    @staticmethod
+    def from_fermion_operator(op: FermionOperator) -> DiagonalCoulombHamiltonian:
+        """Convert a FermionOperator to a DiagonalCoulombHamiltonian."""
+
+        # print("op here = ", op)
+
+        # extract norb
+        orb_list = []
+        for key, _ in op.items():
+            for operator in key:
+                orb = operator[2]
+                orb_list.append(orb)
+        norb = max(orb_list) + 1
+
+        # initialize variables
+        constant: float = 0
+        one_body_tensors = np.zeros((2, norb, norb), dtype=complex)
+        diag_coulomb_mats = np.zeros((3, norb, norb), dtype=float)
+
+        # populate the tensors (new)
+        for term, coeff in op.items():
+            # print(f"(term, coeff) = ({term}, {coeff})")
+            if len(term) == 0:  # constant term
+                constant = np.real(coeff)
+            elif len(term) == 2:  # one-body term candidate
+                p = term[0][2]
+                q = term[1][2]
+                # print(p, q)
+                term_a = (cre_a(p), des_a(q))
+                term_b = (cre_b(p), des_b(q))
+                if term == term_a:
+                    # print("one-body a match")
+                    one_body_tensors[0][p, q] += 0.5 * coeff  # 0.5
+                elif term == term_b:
+                    # print("one-body b match")
+                    one_body_tensors[1][p, q] += 0.5 * coeff  # 0.5
+                else:
+                    raise ValueError(
+                        "FermionOperator cannot be converted to DiagonalCoulombHamiltonian")
+            elif len(term) == 4:  # two-body term candidate
+                p = term[0][2]
+                q = term[2][2]
+                # print(p, q)
+                # two-body terms
+                term_aa = (cre_a(p), des_a(p), cre_a(q), des_a(q))
+                terms_ab = [(cre_a(p), des_a(p), cre_b(q), des_b(q)),
+                            (cre_b(p), des_b(p), cre_a(q), des_a(q))]
+                term_bb = (cre_b(p), des_b(p), cre_b(q), des_b(q))
+                if term == term_aa:
+                    # print("two-body aa match")
+                    diag_coulomb_mats[0][p, q] += 2 * np.real(coeff)
+                elif term in terms_ab:
+                    # print("two-body ab match")
+                    diag_coulomb_mats[1][p, q] += np.real(coeff)
+                elif term == term_bb:
+                    # print("two-body bb match")
+                    diag_coulomb_mats[2][p, q] += 2 * np.real(coeff)
+                else:
+                    raise ValueError(
+                        "FermionOperator cannot be converted to DiagonalCoulombHamiltonian")
+            else:
+                raise ValueError("FermionOperator cannot be converted to DiagonalCoulombHamiltonian")
+
+        # ensure diag_coulomb_mats symmetry
+        for i in range(3):
+            diag_coulomb_mats[i] = (diag_coulomb_mats[i] + diag_coulomb_mats[i].T) / 2
```

Comment:
Extremely minor nit but ffsim convention is to multiply by 0.5 instead of divide by 2. Multiplication is a simpler numerical operation than division and is preferred when the number can be represented exactly in binary.